### PR TITLE
Do not update rx.synchronised when there is no input data

### PR DESCRIPTION
**Pull request by bmerry**
If an accumulation has no data at all, do not do any processing on it,
including updating rx.synchronised. This gives more consistent
behaviour: previously rx.synchronised could get updated if an input
chunk overlapped the accumulation, even if it didn't contain any valid
heaps for that accumulation. With this change, the observable behaviour
no longer depends on the chunk size.

This allows the testing for rx.synchronised to be made more robust
(checking the exact order of updates).

Secondly (in a separate commit), the timestamp associated with rx.timestamp is set from the data. Previously it used wall clock time, which left it at odds with the other
rx.* sensors, which are tied to the data stream timestamps.

This probably opens more opportunities to verify sensor timestamps, but
I'll leave that for another ticket.

Checklist (if not applicable, edit to add `(N/A)` and mark as done):

- [x] (n/a) If dependencies are added/removed: update `setup.cfg` and `.pre-commit-config.yaml`
- [x] (n/a) If modules are added/removed: use `sphinx-apidoc -efo doc/ src/` to update files in `doc/`
- [x] Ensure copyright notices are present and up-to-date
- [x] (n/a) If qualification tests are changed: attach a sample qualification report
- [x] (n/a) If design has changed: ensure documentation is up to date
- [x] (n/a) If ICD-defined sensors have been added: update `fake_servers.py` in katsdpcontroller to match

Closes NGC-1308.

### CODE REVIEW

**Comment by bmerry** [comment]
> This probably opens more opportunities to verify sensor timestamps, but I'll leave that for another ticket.

Specifically, I've opened NGC-1379.

File: test/xbgpu/test_engine.py [hunk 1]
```diff
@@ -335,9 +339,10 @@ def verify_corrprod_sensors(
     *,
     xpipelines: list[XPipeline],
     prom_diff: PromDiff,
-    actual_sensor_updates: dict[str, list[tuple[Any, aiokatcp.Sensor.Status]]],
+    actual_sensor_updates: dict[str, list[aiokatcp.Reading]],
```
Comment (amishatishpatel) [inline test/xbgpu/test_engine.py:342]: Why didn't I use `Reading` in the first place? 🤦

**Review by amishatishpatel — CHANGES_REQUESTED** [review]
One explicit change `TODO`, please. Seeing as you picked up the ticket, see [Line 1033](https://github.com/ska-sa/katgpucbf/blob/NGC-1308-missing-accum/test/xbgpu/test_engine.py#L1033-L1041).

Also, I will have to request you add a `TODO` for NGC-1379 wherever you see fit.

Please and thank yous all around.

File: test/xbgpu/test_engine.py [hunk 2]
```diff
@@ -347,72 +346,72 @@ def verify_corrprod_sensors(
...
-        # but have incomplete data.
+        # but have incomplete data. Also calculate expected updates to
+        # rx.synchronised.
+        heap_accumulation_threshold = xpipeline.output.heap_accumulation_threshold
+        # Note: this truncates any partial final accumulation, because it will
+        # not get flushed.
+        n_accs = len(present) // heap_accumulation_threshold
+        expected_sensor_updates = []
+        complete_accs = 0
         incomplete_accs = 0
```
[the remark this answers is not in the document]
Comment (bmerry) [reply test/xbgpu/test_engine.py:387]: The fact that you felt the need to test things out in ipython suggests that your version is more confusing. I'd rather leave it as is.

**Comment by bmerry** [comment]
> One explicit change TODO, please. Seeing as you picked up the ticket, see [Line 1033](https://github.com/ska-sa/katgpucbf/blob/NGC-1308-missing-accum/test/xbgpu/test_engine.py#L1033-L1041).

Done - thanks for catching.

> Also, I will have to request you add a TODO for NGC-1379 wherever you see fit.

Doing that would require first spending timing on the ticket to figure out which sensor should have their timestamps tested, so I don't think it makes sense to do this.

**Review by amishatishpatel — APPROVED** [review]

**Pull request merged**